### PR TITLE
Auth/signup

### PR DESCRIPTION
**Pull request by festinalente91**
* This is on going pull request, but for the sharing temperary requested.
- updated JWT token
- added generating token, refreshing token.
- added signup
- added authentication. About to use check token validation before use api

### CODE REVIEW

**Comment by hewonjeong** [comment]
Congrats! It's 100th PR~~~

File: src/server/auth-token.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  ACCESS:  process.env.ACCESS_SECRET_KEY || config.ACCESS_SECRET_KEY,
+  REFRESH: process.env.REFRESH_SECRET_KEY || config.REFRESH_SECRET_KEY
+};
+
+/*
+  @TODO
+  for better security, you need to encrypt payload(claims in jwt)
+ */
+class AuthToken {
+
```
Comment (ovekyc) [inline src/server/auth-token.js:30]: Please remove newline

File: src/server/auth/controllers.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import {APIError} from './../ErrorHandler';
+import jwt, {TOKEN_TYPE} from './../auth-token';
+import AuthModel from './models';
+export default {
+  refreshToken: (req , res, next) => {
+    jwt.decode(TOKEN_TYPE.REFRESH, req.body.refreshToken)
+      .then((decoded) => {
```
Comment (ovekyc) [inline src/server/auth/controllers.js]: You can remove one indent if you write .then in same line with jwt.decode()

File: src/server/auth/router.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import express from 'express';
+import controller from './controllers';
+
+const router = express.Router();
+
+
```
Comment (ovekyc) [inline src/server/auth/router.js:36]: Please remove newline

File: src/server/users/controllers.js [hunk 4]
```diff
@@ -69,5 +71,33 @@ export default {
...
+        break;
+      case 'facebook':
+        addUser = UserModel.addFacebookUser(req.body);
+        break;
+      default:
+        break;
+    }
+
+    addUser
+      .then(AuthModel.encodeTokenSet)
```
Comment (ovekyc) [inline src/server/users/controllers.js:124]: ditto

File: src/test/node/auth-token.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+      t.end();
+    })
+    .catch(() => {
+      t.fail('decoded token is not same with payload');
+      t.end();
+    });
+});
+
+test('test authenticate user', t => {
+
```
Comment (ovekyc) [inline src/test/node/auth-token.js]: ditto

File: src/test/node/auth-token.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  const mockUser = {
+    user: 'mockUserKey'
+  };
+  const expected = {
+    userKey: mockUser.user
+  };
+  AuthToken.encode(TOKEN_TYPE.ACCESS, mockUser)
+    .then((accessToken) => {
+
```
Comment (ovekyc) [inline src/test/node/auth-token.js]: ditto

File: src/test/node/auth/controllers.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import jwt, {TOKEN_TYPE} from './../../../server/auth-token';
+import AuthController from '../../../server/auth/controllers';
+import httpMocks from 'node-mocks-http';
+
+test('refresh token', t => {
+  const mockUser = {
+    user: 'mockUserKey'
+  };
+  jwt.encode(TOKEN_TYPE.REFRESH, mockUser)
+    .then((mockRefreshToken) => {
```
Comment (ovekyc) [inline src/test/node/auth/controllers.js]: ditto

File: src/test/node/auth/models.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+      user: mockUserKey
+    },
+    refreshToken: {
+      type: TOKEN_TYPE.REFRESH,
+      user: mockUserKey
+    }
+  };
+
+  AuthModel.encodeTokenSet(mockUserKey)
+    .then(tokenSet => {
```
Comment (ovekyc) [inline src/test/node/auth/models.js:25]: ditto

File: src/test/node/users/controllers.js [hunk 9]
```diff
@@ -92,3 +98,63 @@ test('add a user to database', t => {
     t.end(err);
   });
 });
+
+test('signup as a anonymous user to database', t => {
+
```
Comment (ovekyc) [inline src/test/node/users/controllers.js]: ditto

File: src/test/node/users/controllers.js [hunk 10]
```diff
@@ -92,3 +98,63 @@ test('add a user to database', t => {
...
+  const mockUser = {
+      userType: 'anonymous',
```
Comment (ovekyc) [inline src/test/node/users/controllers.js]: Please stick 2-space indent

File: src/test/node/users/controllers.js [hunk 11]
```diff
@@ -92,3 +98,63 @@ test('add a user to database', t => {
...
+  const req = httpMocks.createRequest({
+    method: 'POST',
+    url: '/users/signup',
+    body: mockUser
+  });
+
+  const res = httpMocks.createResponse();
+
+  clearDB()
+    .then(() => {
```
Comment (ovekyc) [inline src/test/node/users/controllers.js:120]: ditto

File: src/test/node/users/controllers.js [hunk 12]
```diff
@@ -92,3 +98,63 @@ test('add a user to database', t => {
...
+
+  const req = httpMocks.createRequest({
+    method: 'POST',
+    url: '/users/signup',
+    body: mockUser
+  });
+  const res = httpMocks.createResponse();
+
+  clearDB()
+    .then(() => {
```
Comment (ovekyc) [inline src/test/node/users/controllers.js]: ditto

File: src/test/node/users/models.js [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+  const mockUser = {
+    key : UserModel.genUserKey(),
+    secret: 'userSecret'
+  };
+  const expected = {
+    value: mockUser
+  };
+
+  clearDB()
+    .then(() => {
```
Comment (ovekyc) [inline src/test/node/users/models.js]: ditto

File: src/test/node/users/models.js [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+  const mockUser = {
+    secret: 'userSecret'
+  };
+  const expected = {
+    type: USER_TYPE.ANONYMOUS,
+    value: mockUser
+  };
+
+  clearDB()
+    .then(() => {
```
Comment (ovekyc) [inline src/test/node/users/models.js]: ditto

File: src/test/node/users/models.js [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+
+test('add Facebook user', t => {
+  const mockUser = {
+    facebookToken: 'facebookToken'
+  };
+  const expected = {
+    type: USER_TYPE.FACEBOOK,
+    value: mockUser
+  };
+  clearDB()
```
Comment (ovekyc) [inline src/test/node/users/models.js:70]: ditto

**Comment by ovekyc** [comment]
@festinalente91 Please check out lint warning

File: src/server/auth/controllers.js (hunk 2 above)
Comment (hewonjeong) [reply src/server/auth/controllers.js]: or please remove one indent at this line.
is there no lint warning because of indent?

**Comment by hewonjeong** [comment]
LGTM. pls merge after fixing lint warnig~!

File: src/server/auth/router.js [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+ *
+ * @apiError (Error 400) error the refresh token is invalid.
+ * @apiErrorExample {json} Invalid-Refresh-Token-Response:
+ *     HTTP/1.1 400 Bad Request
+ *     {
+ *       error: "error message ..."
+ *     }
+ *
+ */
+router.post('/auth/refresh', controller.refreshToken);
```
Comment (ovekyc) [inline src/server/auth/router.js]: I think the address has to be changed 
Please check out server/router/api-routes.js & server/<module>/router.js
OR check out server/index.js

File: src/server/auth/router.js [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+ *     }
+ *
+ * @apiError (Error 500) error database error
+ * @apiErrorExample {json} Database-Error-Response:
+ *     HTTP/1.1 500 Internal Server Error
+ *     {
+ *       error: "error message ..."
+ *     }
+ */
+router.post('/auth/grant', controller.grant);
```
Comment (ovekyc) [inline src/server/auth/router.js]: ditto

File: src/server/users/controllers.js [hunk 18]
```diff
@@ -97,5 +100,52 @@ export default {
       });
       return cb();
     });
+  },
+  signup: (req, res, next) => {
+    const {userType} = req.body;
+    let addUser;
+
+    switch (userType) {
+    case 'anonymous':
```
Comment (ovekyc) [inline src/server/users/controllers.js]: You can define 'anonymous', 'facebook' as constant

File: src/test/node/auth/controllers.js [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+          });
+      });
+    })
+    .catch(err => {
+      t.fail();
+      t.end(err);
+    });
+});
+
+// grant: (req, res, next) => {
```
Comment (ovekyc) [inline src/test/node/auth/controllers.js]: please remove this comments

**Comment by ovekyc** [comment]
Too big chunk.

**Comment by ovekyc** [comment]
I worried about the coverage of test code
I didnt see the test code of bycryt, facebookmanager and usermanager
Please make sure that will be fine

**Comment by festinalente91** [comment]
Be careful when you use process.env and config.variable.
In some library ( like this bcrypt), some variable must be a number type.
However, if you get variable from process.env, the variable's type is a string.
As a result, you might pass the test in local using config file, but it cannot be passed in deployment environment using process.env.

[https://www.bountysource.com/issues/36106806-wrong-error-message-being-thrown-when-salt-rounds-value-is-a-string](https://www.bountysource.com/issues/36106806-wrong-error-message-being-thrown-when-salt-rounds-value-is-a-string)

**Comment by festinalente91** [comment]
also... when you config travis process.env variable, if the variable is more than a word
(which contains spacebars), then you should add double quotes.

File: src/server/users/controllers.js [hunk 20]
```diff
@@ -1,8 +1,11 @@
-import { KeyUtils, ENTITY, STATE } from '../key-utils';
-import { S3Connector } from '../aws-s3';
-import db from '../database';
-import { APIError } from '../ErrorHandler';
-import { CreatedPostManager, SavedPostManager } from './models';
+import {KeyUtils, ENTITY, STATE} from '../key-utils';
+import {S3Connector} from '../aws-s3';
+import db from './../database';
```
Comment (ovekyc) [inline src/server/users/controllers.js]: ./ is meaningless

File: src/server/users/controllers.js [hunk 21]
```diff
@@ -1,8 +1,11 @@
...
+import {APIError} from './../ErrorHandler';
```
Comment (ovekyc) [inline src/server/users/controllers.js]: ditto

File: src/server/users/controllers.js [hunk 22]
```diff
@@ -1,8 +1,11 @@
...
+import AuthModel from './../auth/models';
```
Comment (ovekyc) [inline src/server/users/controllers.js]: ditto

File: src/server/users/controllers.js [hunk 23]
```diff
@@ -97,5 +100,47 @@ export default {
...
+    switch (userType) {
+    case 'anonymous':
+      req.body.secret = uuid();
+      addUser = UserModel.addAnonymousUser(req.body);
+      break;
+    case 'facebook':
+      addUser = UserModel.addFacebookUser(req.body);
+      break;
+    default:
+      break;
```
Comment (ovekyc) [inline src/server/users/controllers.js]: You should call next function with Error

Comment (ovekyc) [reply src/server/users/controllers.js]: break will raise exception

File: src/server/users/models.js [hunk 24]
```diff
@@ -1,8 +1,155 @@
+import {KeyUtils, ENTITY, STATE} from './../key-utils';
```
Comment (ovekyc) [inline src/server/users/models.js]: ditto

File: src/server/users/models.js [hunk 25]
```diff
@@ -1,8 +1,155 @@
...
+import bcrypt from './../bcrypt';
```
Comment (ovekyc) [inline src/server/users/models.js]: ditto

File: src/server/users/controllers.js [hunk 26]
```diff
@@ -97,5 +100,47 @@ export default {
...
+    }
+    let userSecret;
+    addUser
+    .then((userKey) => {
+      switch (userType) {
```
Comment (ovekyc) [inline src/server/users/controllers.js]: this switch statement is not needed. You can deal userSecret in
switch statement at line 107 of this file.

File: src/server/users/models.js [hunk 27]
```diff
@@ -1,8 +1,155 @@
...
+  }
+};
+
+export const USER_TYPE = {
+  ANONYMOUS: 0,
+  FACEBOOK: 1
+};
+
+const UserManager = {
+  addUser: (key, value) => {
```
Comment (ovekyc) [inline src/server/users/models.js]: This function is just promise wrapping of db.put
Please use just db.put or
Add function to db directly something like 'putPromise'

File: src/server/users/models.js [hunk 28]
```diff
@@ -1,8 +1,155 @@
...
+    return new Promise((resolve, reject) => {
+      db.put(key, value, (err) => {
+        if (err) {
+          return reject(err);
+        }
+        return resolve(key);
+      });
+    });
+  },
+  getUser: (key) => {
```
Comment (ovekyc) [inline src/server/users/models.js]: This function too.
Please refine your code.

File: src/server/users/models.js [hunk 29]
```diff
@@ -1,8 +1,155 @@
...
+      .then(userData => {
+        return userData.key;
+      })
+  }
+  ,
+  addAnonymousUser: (data) => {
+    const userKey = UserManager.genUserKey();
+    const userValue = {
+      type: USER_TYPE.ANONYMOUS,
+      name: 'guest',
```
Comment (ovekyc) [inline src/server/users/models.js]: How about define GUEST constant?

File: src/server/users/models.js [hunk 30]
```diff
@@ -1,8 +1,155 @@
...
+          secret: data.secret
+        });
+        return UserManager.addUser(userIdx, {key});
+      });
+  },
+  addFacebookUser: (data) => {
+    /*
+      This method is called when facebook user sign up.
+      It save user data(name, email, facebook) from facebook graph API.
+      Also saved in db as 'user-timehash' and index like 'user-0-facebook-0000000(facebookId)'
```
Comment (ovekyc) [inline src/server/users/models.js]: Also save

**Comment by ovekyc** [comment]
Review will be continued from src/test/node/auth-token.js ....

File: src/test/node/auth/controllers.js [hunk 31]
```diff
@@ -0,0 +1,220 @@
+import test from 'tape';
+import AuthToken, {TOKEN_TYPE} from './../../../server/auth-token';
```
Comment (ovekyc) [inline src/test/node/auth/controllers.js:2]: ./ is meaningless

**Comment by hewonjeong** [comment]
[As a server administrator, I'd like to correspond data with verified user so that I can protect information](https://trello.com/c/iL3ZBZCQ/170-as-a-server-administrator-i-d-like-to-correspond-data-with-verified-user-so-that-i-can-protect-information)

File: src/test/node/auth-token.js [hunk 32]
```diff
@@ -19,3 +22,31 @@ test('test auto token instance', t => {
...
+    user: 'mockUserKey'
+  };
+  const expected = {
+    userKey: mockUser.user
+  };
+  AuthToken.encode(TOKEN_TYPE.ACCESS, mockUser)
+    .then((accessToken) => {
+      const req = httpMocks.createRequest({
+        method: 'POST',
+        url: '/users',
```
Comment (ovekyc) [inline src/test/node/auth-token.js]: api/users maybe

File: src/test/node/auth/controllers.js [hunk 33]
```diff
@@ -0,0 +1,220 @@
...
+          user: mockUser.user
+        },
+        refreshToken: {
+          type: TOKEN_TYPE.REFRESH,
+          user: mockUser.user
+        }
+      };
+      const req = httpMocks.createRequest({
+        method: 'POST',
+        url: '/users',
```
Comment (ovekyc) [inline src/test/node/auth/controllers.js]: api/users

File: src/test/node/auth/controllers.js [hunk 34]
```diff
@@ -0,0 +1,220 @@
...
+    }
+  };
+
+  bcrypt.hash(mockUserSecret)
+    .then(hash => {
+      mockUser.secret = hash;
+    })
+    .then(clearDB)
+    .then(() => {
+      return new Promise((resolve, reject) => {
```
Comment (ovekyc) [inline src/test/node/auth/controllers.js]: If you define putPromise function as i suggested
You can edit this code too

File: src/test/node/auth/controllers.js [hunk 35]
```diff
@@ -0,0 +1,220 @@
...
+        db.put(mockUser.key, mockUser, err => {
+          if (err) {
+            return reject(err);
+          }
+          return resolve();
+        });
+      });
+    })
+    .then(() => {
+      return new Promise((resolve, reject) => {
```
Comment (ovekyc) [inline src/test/node/auth/controllers.js]: And same here (line 85)

File: src/test/node/auth/controllers.js [hunk 36]
```diff
@@ -0,0 +1,220 @@
...
+            return reject(err);
+          }
+          return resolve();
+        });
+      });
+    })
+    .then(() => {
+      const req = httpMocks.createRequest({
+        method: 'POST',
+        url: '/auth/grant',
```
Comment (ovekyc) [inline src/test/node/auth/controllers.js]: /api/auth/grant/

File: src/test/node/auth/controllers.js [hunk 37]
```diff
@@ -0,0 +1,220 @@
...
+      t.end(err);
+    });
+});
+test('grant facebook user in controller', t => {
+  const mockUser = {
+    key: 'userKey',
+    type: USER_TYPE.FACEBOOK,
+    facebookId: process.env.FACEBOOK_TEST_ID || config.FACEBOOK_TEST_ID
+  };
+  const mockUserIdxKey = `${ENTITY.USER}-${STATE.ALIVE}` +
```
Comment (ovekyc) [inline src/test/node/auth/controllers.js:138]: Why use + ? 
I think it can written using only template

File: src/test/node/auth/controllers.js [hunk 38]
```diff
@@ -0,0 +1,220 @@
...
+      user: mockUser.key
+    },
+    refreshToken: {
+      type: TOKEN_TYPE.REFRESH,
+      user: mockUser.key
+    }
+  };
+  clearDB()
+    .then(() => {
+      return new Promise((resolve, reject) => {
```
Comment (ovekyc) [inline src/test/node/auth/controllers.js]: if you define putPromise,,, 
same here

File: src/test/node/auth/controllers.js [hunk 39]
```diff
@@ -0,0 +1,220 @@
...
+          if (err) {
+            return reject(err);
+          }
+          return resolve();
+        });
+      });
+    })
+    .then(() => {
+      return new Promise((resolve, reject) => {
+        db.put(mockUserIdxKey, {key: mockUser.key}, err => {
```
Comment (ovekyc) [inline src/test/node/auth/controllers.js]: same here

File: src/test/node/auth/controllers.js [hunk 40]
```diff
@@ -0,0 +1,220 @@
...
+          }
+          return resolve();
+        });
+      });
+    })
+    .then(FacebookManager.getTestAccessToken)
+    .then(mockFacebookToken => {
+      const req = httpMocks.createRequest({
+        method: 'POST',
+        url: '/auth/grant',
```
Comment (ovekyc) [inline src/test/node/auth/controllers.js]: /api/auth/grant

File: src/test/node/auth/models.js [hunk 41]
```diff
@@ -0,0 +1,124 @@
+import test from 'tape';
+import jwt, {TOKEN_TYPE} from './../../../server/auth-token';
```
Comment (ovekyc) [inline src/test/node/auth/models.js]: ./ is meaningless

File: src/test/node/auth/models.js [hunk 42]
```diff
@@ -0,0 +1,124 @@
...
+  const mockUserIdxKey = `${ENTITY.USER}-${STATE.ALIVE}-${ENTITY.ANONYMOUS}-${mockUserSecret}`;
+
+  bcrypt.hash(mockUserSecret)
+    .then(hash => {
+      mockUser.secret = hash;
+    })
+    .then(clearDB)
+    .then(() => {
+      return new Promise((resolve, reject) => {
+        db.put(mockUser.key, mockUser, err => {
```
Comment (ovekyc) [inline src/test/node/auth/models.js]: same here

File: src/test/node/auth/models.js [hunk 43]
```diff
@@ -0,0 +1,124 @@
...
+          if (err) {
+            return reject(err);
+          }
+          return resolve();
+        });
+      });
+    })
+    .then(() => {
+      return new Promise((resolve, reject) => {
+        db.put(mockUserIdxKey, {key: mockUser.key}, err => {
```
Comment (ovekyc) [inline src/test/node/auth/models.js]: same here

File: src/test/node/auth/models.js [hunk 44]
```diff
@@ -0,0 +1,124 @@
...
+    type: USER_TYPE.FACEBOOK,
+    facebookId: process.env.FACEBOOK_TEST_ID || config.FACEBOOK_TEST_ID
+  };
+  const mockUserIdxKey = `${ENTITY.USER}-${STATE.ALIVE}` +
+    `-${ENTITY.FACEBOOK}-${mockUser.facebookId}`;
+
+  clearDB()
+    .then(() => {
+      return new Promise((resolve, reject) => {
+        db.put(mockUser.key, mockUser, err => {
```
Comment (ovekyc) [inline src/test/node/auth/models.js]: Same here

File: src/test/node/auth/models.js [hunk 45]
```diff
@@ -0,0 +1,124 @@
...
+          if (err) {
+            return reject(err);
+          }
+          return resolve();
+        });
+      });
+    })
+    .then(() => {
+      return new Promise((resolve, reject) => {
+        db.put(mockUserIdxKey, {key: mockUser.key}, err => {
```
Comment (ovekyc) [inline src/test/node/auth/models.js]: same here

File: src/server/users/controllers.js (hunk 23 above)
Comment (festinalente91) [reply src/server/users/controllers.js]: good point!

File: src/server/users/models.js (hunk 30 above)
Comment (festinalente91) [reply src/server/users/models.js]: always. @ovekyc - compiler has amazing performance.

File: src/test/node/auth/controllers.js (hunk 37 above)
Comment (festinalente91) [reply src/test/node/auth/controllers.js:138]: @ovekyc special thanks to lint error :)

**Comment by hewonjeong** [comment]
sdjidjafdsjfjttttapm'f[p'ds,fmrlgtmsdpfjajfdsjfjsd;afdsf;sdkfd;s

File: src/server/database.js [hunk 46]
```diff
@@ -19,6 +19,28 @@ export const clearDB = () => {
   });
 };
 
+export const putPromise = (key, value) => {
+  return new Promise((resolve, reject) => {
```
Comment (hewonjeong) [inline src/server/database.js:23]: Good! we really needed these.
but how bout just 'put' or 'get'?
Cause we dont named with 'promise' if those functions return Promise object.

Comment (festinalente91) [reply src/server/database.js:23]: Right. However there are many lagacy codes that are using levelup put and get method, so  to reduce confusion I added the "Promise" postfix. Later merging current PRs, let's talk about adapting promise in db :)

File: src/server/auth-token.js [hunk 47]
```diff
@@ -0,0 +1,68 @@
...
+  constructor() {}
+
+  encode(type, payload) {
+    return new Promise((resolve) => {
+      payload.type = type;
+      jwt.sign(payload, SECRET_KEY[type], { expiresIn: TOKEN_EXPIRE[type]}, (err, token) => {
+        if (err) {
+          throw err;
```
Comment (hewonjeong) [inline src/server/auth-token.js]: Why it throw err object? not reject?

File: src/server/users/models.js [hunk 48]
```diff
@@ -0,0 +1,49 @@
...
+  addUser: (key, value) => {
+    return new Promise((resolve, reject) => {
+      db.put(key, value, (err) => {
+        if (err) {
+          return reject(err);
+        }
+        return resolve(key);
+      });
+    });
+  },
```
Comment (hewonjeong) [inline src/server/users/models.js:21]: How about just 'add' cause it's UserModel?

**Comment by hewonjeong** [comment]
LGTM

**Comment by ovekyc** [comment]
LGTM

**Comment by hewonjeong** [comment]
lgtm again

**Pull request merged**